[JSC] Support JSVALUE32_64 in WasmIPIntSlowPaths

eugeneia opened the pull request:
#### c72d01c78f73cc6d3e3c4a47f8c8afce2908d398
<pre>
[JSC] Support JSVALUE32_64 in WasmIPIntSlowPaths
<a href="https://bugs.webkit.org/show_bug.cgi?id=291741">https://bugs.webkit.org/show_bug.cgi?id=291741</a>

Reviewed by Keith Miller.

We return from IPInt slow path operations as follows:

 - when we return an exception:
   + on 64 bit: r0(SlowPathExceptionTag=1) r1(exception)
   + on 32 bit: r0(SlowPathExceptionTag=InvalidTag) r1(exception)
 - when we return a JSValue
   + on 64 bit: r0(EncodedJSValue) r1(0)
   + on 32 bit: r0(EncodedJSValue/Tag) r1(EncodedJSValue/Payload)

Also fixes some floating point instructions I came across by chance
in InPlaceInterpreter64.asm

* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/llint/LLIntOffsetsExtractor.cpp:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL_1P):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.h:

Canonical link: <a href="https://commits.webkit.org/293913@main">https://commits.webkit.org/293913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7b8fdce58fcf98cc7b4a9b3346a10fdf8ed8317c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100298 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19947 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10245 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105428 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50882 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28421 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/76371 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/33430 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103305 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15503 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90594 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56727 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15324 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8594 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50249 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/92956 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85236 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8675 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107787 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/98905 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20100 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/85324 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27774 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86801 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84862 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21558 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29543 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/7277 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21321 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27348 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32581 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/122531 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27159 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34194 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30475 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28718 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
```diff
@@ -73,14 +73,20 @@ namespace JSC { namespace IPInt {
+#define WASM_RETURN_JSVALUE(value) WASM_RETURN_TWO(std::bit_cast<void*>(JSValue::decode(value).tag()), std::bit_cast<void*>(JSValue::decode(value).payload()))
+#endif
+
 // For operation calls that may throw an exception, we return (0, <val>)
```
Can we add a FIXME to have this match what JS does on 64-bit? i.e.` return (<exception>, 0)`

eugeneia replied:
Oh yes, i.e. flip the order? That would indeed be neat! I'm not going to get to it anymore this week, but I might just do this next week because I noticed that its also inconsistent with how we deal with register pairs on JSVALUE32_64

kmiller68 reviewed (approved):
r=me

danlliu commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -2816,16 +2816,12 @@ instructionLabel(_i64_reinterpret_f64)
 instructionLabel(_f32_reinterpret_i32)
-    pushInt32(t0)
-    fi2f t0, ft0
-    popFloat32(ft0)
```
How do these end up as a `nop`? Shouldn't we be changing the format from 2's complement to IEEE-754?

eugeneia replied:
The deleted code obviously is bogus and a NOP as is, right? (push garbage -> pop garbage)

And its a nop on little-endian as far as I understand because this is supposed to reinterpret 32s bits as a IEEE float, which is what we do automatically when we `popFloat32`.

eugeneia replied:
There is also f32.convert_i32_s/u which would convert an int to a float which do `fi2f` AFAIU.

danlliu commented on Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
```diff
@@ -2816,16 +2816,12 @@ instructionLabel(_i64_reinterpret_f64)
 instructionLabel(_f64_reinterpret_i64)
-    pushInt64(t0)
-    fq2d t0, ft0
-    popFloat64(ft0)
```
(same here)

eugeneia committed:
[JSC] Support JSVALUE32_64 in WasmIPIntSlowPaths
https://bugs.webkit.org/show_bug.cgi?id=291741

Reviewed by Keith Miller.

We return from IPInt slow path operations as follows:

 - when we return an exception:
   + on 64 bit: r0(SlowPathExceptionTag=1) r1(exception)
   + on 32 bit: r0(SlowPathExceptionTag=InvalidTag) r1(exception)
 - when we return a JSValue
   + on 64 bit: r0(EncodedJSValue) r1(0)
   + on 32 bit: r0(EncodedJSValue/Tag) r1(EncodedJSValue/Payload)

Also fixes some floating point instructions I came across by chance
in InPlaceInterpreter64.asm

* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/llint/LLIntOffsetsExtractor.cpp:
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.cpp:
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL):
(JSC::IPInt::WASM_IPINT_EXTERN_CPP_DECL_1P):
* Source/JavaScriptCore/wasm/WasmIPIntSlowPaths.h:

Canonical link: https://commits.webkit.org/293913@main
